### PR TITLE
Refactor Native /Assembly resolution

### PR DESCRIPTION
**Pull request by KevinRansom**
This moves native resolution from CompilerOps to the InteractiveDependencyManager.dll
It adds an assembly resolve handler for scripting scenarios.

Note:  *** Still need to figure out what to do about ***   
1.   CreatePackageManagerUnknownError
2.   reportError:(ErrorReportType -> int * string -> unit) in
      * TryFindDependencyManagerByKey
      * TryFindDependencyManagerInPath

### CODE REVIEW

File: src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.Utilities.fs [hunk 1]
```diff
@@ -152,16 +152,16 @@ module internal Utilities =
...
+        let mutable list = List.empty
         let rec copyLines () =
             match stream.ReadLine() with
             | null -> ()
             | line ->
-                writer.WriteLine(line)
+                list <- line :: list
                 copyLines ()
         copyLines ()
+        list |> List.toArray
```
Comment (brettfo) [inline src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.Utilities.fs]: `|> Array.rev` since the list is created backwards.

File: src/fsharp/CompileOps.fs [hunk 2]
```diff
@@ -2390,9 +2390,24 @@ type TcConfigBuilder =
           noConditionalErasure = false
           pathMap = PathMap.empty
           langVersion = LanguageVersion("default")
-          dependencyProvider = new DependencyProvider()
+          dependencyProvider = Unchecked.defaultof<DependencyProvider>
```
Comment (cartermp) [inline src/fsharp/CompileOps.fs]: Why this change?

File: src/fsharp/CompileOps.fs [hunk 3]
```diff
@@ -2390,9 +2390,24 @@ type TcConfigBuilder =
...
+    //  2. compilerToolPath directories
+    //  3. reference dll's
+    //  4. The implicit include directory
+    member private tcConfigB.nativeProbingRoots () =
+        seq {
+            yield! tcConfigB.includes
+            yield! tcConfigB.compilerToolPaths
+            yield! (tcConfigB.referencedDLLs |> Seq.map(fun ref -> Path.GetDirectoryName(ref.Text)))
+            yield tcConfigB.implicitIncludeDir
+        } |>Seq.distinct
```
Comment (cartermp) [inline src/fsharp/CompileOps.fs]:
```suggestion
        } |> Seq.distinct
```

File: src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.Utilities.fs (hunk 1 above)
Comment (cartermp) [inline src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.Utilities.fs]: I think a `ResizeArray` would be preferred here since you seem to be effectively copying the semantics of one but with the F# list type

File: src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.fs [hunk 4]
```diff
@@ -61,11 +61,11 @@ module FSharpDependencyManager =
                     let setVersion v = Some { current with Version = v }
                     match opt with
                     | Some "include", Some v -> addInclude v |> parsePackageReferenceOption' rest implicitArgumentCount
-                    | Some "include", None -> raise (ArgumentException(sprintf "%s requires a value" "Include"))               // @@@@@@@@@@@@@@@@@@@@@@@ Globalize me please
+                    | Some "include", None -> raise (ArgumentException(sprintf "%s requires a value" "Include"))                    // @@@@@@@@@@@@@@@@@@@@@@@ Globalize me please
```
Comment (cartermp) [inline src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.fs]: > Globalize me please

Are you intending on plumbing through a string from a settings file given comments like this?

File: src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.fs [hunk 5]
```diff
@@ -136,7 +140,7 @@ type [<DependencyManagerAttribute>] FSharpDependencyManager (outputDir:string op
 
     member __.Key = key
 
-    member __.ResolveDependencies(scriptExt:string, packageManagerTextLines:string seq, tfm: string) : bool * string seq * string seq * string seq =
+    member __.ResolveDependencies(scriptExt:string, packageManagerTextLines:string seq, tfm: string) : bool * string array * string array * string seq * string seq * string seq =
```
Comment (cartermp) [inline src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.fs]: Giving me anxiety here Kevin 🙂 

How do you feel about pulling this out into a record? A 6-tuple feels like a bit much

File: src/fsharp/Interactive.DependencyManager/AssemblyResolveHandler.fs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+open System.IO
+open System.Reflection
+
+#if NETSTANDARD
+open System.Runtime.Loader
+#endif
+
+type AssemblyResolutionProbe = delegate of Unit -> IEnumerable<string>
+
+type AssemblyResolveHandler  (assemblyProbingPaths: AssemblyResolutionProbe) =
```
Comment (cartermp) [inline src/fsharp/Interactive.DependencyManager/AssemblyResolveHandler.fs]:
```suggestion
type AssemblyResolveHandler (assemblyProbingPaths: AssemblyResolutionProbe) =
```

File: src/fsharp/Interactive.DependencyManager/DependencyProvider.fs [hunk 7]
```diff
@@ -58,18 +60,45 @@ module ReflectionHelper =
...
+    /// The resolution paths
+    member public __.Resolutions = resolutions
+
+    /// The source code file paths
+    member public __.SourceFiles = sourceFiles
+
+    /// The roots to package directories
+    member public __.Roots = roots
+
+
```
Comment (cartermp) [inline src/fsharp/Interactive.DependencyManager/DependencyProvider.fs]:
```suggestion

```

File: src/fsharp/Interactive.DependencyManager/DependencyProvider.fs [hunk 8]
```diff
@@ -14,8 +14,10 @@ module ReflectionHelper =
     let resolveDependenciesMethodName = "ResolveDependencies"
     let namePropertyName = "Name"
     let keyPropertyName = "Key"
+    let arrEmpty = Array.empty<string>
     let seqEmpty = Seq.empty<string>
 
+
```
Comment (cartermp) [inline src/fsharp/Interactive.DependencyManager/DependencyProvider.fs]: What is this extra line?

File: src/fsharp/Interactive.DependencyManager/DependencyProvider.fs [hunk 9]
```diff
@@ -115,27 +144,29 @@ type ReflectionDependencyManagerProvider(theType: Type, nameProperty: PropertyIn
...
                 //     3 - (bool * string list * string list)
                 let tupleFields = result |> FSharpValue.GetTupleFields
                 match tupleFields |> Array.length with
-                | 4 -> tupleFields.[0] :?> bool, tupleFields.[1] :?> string seq, tupleFields.[2] :?> string seq, tupleFields.[3] :?> string seq
-                | 3 -> tupleFields.[0] :?> bool, seqEmpty, tupleFields.[1] :?> string list  |> List.toSeq, tupleFields.[2] :?> string list |> List.toSeq
-                | _ -> false, seqEmpty, seqEmpty, seqEmpty
-
-            succeeded, references, generatedScripts, additionalIncludeFolders
+                | 6 -> tupleFields.[0] :?> bool, tupleFields.[1] :?> string array, tupleFields.[2] :?> string array, tupleFields.[3] :?> string seq, tupleFields.[4] :?> string seq, tupleFields.[5] :?> string seq
+                | 3 -> tupleFields.[0] :?> bool, arrEmpty, arrEmpty, seqEmpty, tupleFields.[1] :?> string list  |> List.toSeq, tupleFields.[2] :?> string list |> List.toSeq
```
Comment (cartermp) [inline src/fsharp/Interactive.DependencyManager/DependencyProvider.fs]: Seeing this makes me really want it to be a record with nice names.

File: src/fsharp/Interactive.DependencyManager/DependencyProvider.fs [hunk 10]
```diff
@@ -215,16 +251,17 @@ type DependencyProvider () =
...
         | e ->
             let e = stripTieWrapper e
-            reportError ErrorReportType.Error (InteractiveDependencyManager.SR.packageManagerError(e.Message))
+            let err, msg = InteractiveDependencyManager.SR.packageManagerError(e.Message)
+            reportError.Invoke(ErrorReportType.Error, err, msg)
             null, Unchecked.defaultof<IDependencyManagerProvider>
 
     /// Remove the dependency mager with the specified key
-    member _.RemoveDependencyManagerKey(packageManagerKey:string, path:string): string =
+    member __.RemoveDependencyManagerKey(packageManagerKey:string, path:string): string =
```
Comment (cartermp) [inline src/fsharp/Interactive.DependencyManager/DependencyProvider.fs]: Why the double `_`?

File: src/fsharp/Interactive.DependencyManager/foo.sig [hunk 11]
```diff
@@ -0,0 +1,188 @@
+#light
```
Comment (cartermp) [inline src/fsharp/Interactive.DependencyManager/foo.sig]: (Putting this at the top)

What is the significance of the term `foo` for the name of this file?

File: src/fsharp/Interactive.DependencyManager/NativeDllResolveHandler.fs [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+                            probingRids |> Seq.tryPick(fun rid -> probeForNativeLibrary root rid name)))
+
+        match probe with
+        | Some path -> NativeAssemblyLoadContext.NativeLoadContext.LoadNativeLibrary(path)
+        | None -> IntPtr.Zero
+
+    // netstandard 2.1 has this property, unfortunately we don't build with that yet
+    //public event Func<Assembly, string, IntPtr> ResolvingUnmanagedDll
+    let eventInfo = typeof<AssemblyLoadContext>.GetEvent("ResolvingUnmanagedDll")
+    let handler = Func<System.Reflection.Assembly, string, IntPtr> (_resolveUnmanagedDll)
```
Comment (cartermp) [inline src/fsharp/Interactive.DependencyManager/NativeDllResolveHandler.fs]:
```suggestion
    let handler = Func<Assembly, string, IntPtr> (_resolveUnmanagedDll)
```

Since you have `System.Reflection` already open.

**Comment by KevinRansom** [comment]
@cartermp foo.sig was a file that shouldn't have been committed.
the double underscore is necessary because the fcs build depends on an F# compiler that doesn't support _. for members.

File: src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.fs (hunk 5 above)
Comment (smoothdeveloper) [reply src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.fs]: @cartermp I think this is not possible because it requires to reference the record type from third party extensions.

This infrastructure is working via runtime reflection and should reference only BCL stuff AFAIU.

**Comment by KevinRansom** [comment]
@cartermp, as for the 6 tuple, yeah, it's ugly as sin.  We need a better mechanism than we have, but this is what we use at the moment.  Perhaps a C# tuple decorated with names.  The nicest solution would have a proper strongly typed API, however, there is not really a good assembly to hold the type definition, so that would involve deploying a new assembly.

2 alternatives I am considering is to return a dictionary <string, obj>  I.e. a property bag.  Alternatively, a reflectable type, where we use reflection to get the values from properties on the type.

The existing mechanism, is ugly as sin, and I would like to do better, but I think we need to ponder it for a bit first.

Kevin

File: src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.fs (hunk 5 above)
Comment (KevinRansom) [reply src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.fs]: @smoothdeveloper, we can do something, it's just code after all.  It's just a matter of "will it be an improvement or not"?.

I think there are 2 decent'ish options:
1. A type with properties that are reflectable.
2. A Dictionary<string, obj>()

In both cases we can easily wrap it to make it strongly typed and usable in DependencyProvider, so there would be less ugly stuff.

Comment (cartermp) [reply src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.fs]: Anonymous record?

File: src/fsharp/Interactive.DependencyManager/Interactive.DependencyManager.fsproj [hunk 13]
```diff
@@ -49,6 +55,7 @@
     <PackageReference Include="System.Reflection.TypeExtensions" Version="$(SystemReflectionTypeExtensionsVersion)" />
     <PackageReference Include="System.Runtime" Version="$(SystemRuntimeVersion)" />
     <PackageReference Include="System.Runtime.InteropServices" Version="$(SystemRuntimeInteropServicesVersion)" />
+    <PackageReference Include="System.Runtime.Loader" Version="$(SystemRuntimeLoaderVersion)" />
```
Comment (auduchinok) [inline src/fsharp/Interactive.DependencyManager/Interactive.DependencyManager.fsproj]: This breaks when using `netstandard2.0` FCS build in .NET Framework process.

Comment (cartermp) [reply src/fsharp/Interactive.DependencyManager/Interactive.DependencyManager.fsproj]: cc @KevinRansom - we'll need to check if this is inherent to targeting net461 or a separate issue

**Comment by KevinRansom** [comment]
@auduchinok,
netstandard2.0 FCS has for at least two years had a reference to:
System.Runtime.Loader [here](https://github.com/dotnet/fsharp/blame/master/fcs/FSharp.Compiler.Service/FSharp.Compiler.Service.fsproj#L718)

I assume it's the specific version that is a problem, I will take a look.

Kevin

**Comment by auduchinok** [comment]
OK, it may be something different with using a type from it then.
Reverting this commit makes it work again, otherwise it fails at runtime with exception. I'll look into it further for details.

**Comment by KevinRansom** [comment]
@auduchinok, actually now that I think about, this assembly isn't part of FCS the code is directly compiled in.  Let me dive in a bit, I think there may be an issue running the netstandard2.0 on the desktop.
